src/taskgraph/util/schema.py

```diff
@@ -2,67 +2,109 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
-
 import pprint
 import re
 from collections.abc import Mapping
+from functools import reduce
+from typing import Literal, Optional, Union
 
+import msgspec
 import voluptuous
 
 import taskgraph
 from taskgraph.util.keyed_by import evaluate_keyed_by, iter_dot_path
 
+# Common type definitions that are used across multiple schemas
+TaskPriority = Literal[
+    "highest", "very-high", "high", "medium", "low", "very-low", "lowest"
+]
+
 
 def validate_schema(schema, obj, msg_prefix):
     """
     Validate that object satisfies schema.  If not, generate a useful exception
     beginning with msg_prefix.
+
+    Args:
+        schema: A voluptuous.Schema or msgspec-based StructSchema type
+        obj: Object to validate
+        msg_prefix: Prefix for error messages
     """
     if taskgraph.fast:
         return
+
     try:
-        schema(obj)
-    except voluptuous.MultipleInvalid as exc:
-        msg = [msg_prefix]
-        for error in exc.errors:
-            msg.append(str(error))
-        raise Exception("\n".join(msg) + "\n" + pprint.pformat(obj))
+        # Handle voluptuous Schema
+        if isinstance(schema, voluptuous.Schema):
+            schema(obj)
+        # Handle msgspec-based schemas (StructSchema and its subclasses)
+        elif isinstance(schema, type) and issubclass(schema, msgspec.Struct):
+            # Check if it's our Struct subclass with validate method
+            if issubclass(schema, Struct):
+                schema.validate(obj)
+            else:
+                # Fall back to msgspec.convert for validation
+                msgspec.convert(obj, schema)
+        else:
+            raise TypeError(f"Unsupported schema type: {type(schema)}")
+    except (
+        voluptuous.MultipleInvalid,
+        msgspec.ValidationError,
+        msgspec.DecodeError,
+    ) as exc:
+        if isinstance(exc, voluptuous.MultipleInvalid):
+            msg = [msg_prefix]
+            for error in exc.errors:
+                msg.append(str(error))
+            raise Exception("\n".join(msg) + "\n" + pprint.pformat(obj))
+        else:
+            raise Exception(f"{msg_prefix}\n{str(exc)}\n{pprint.pformat(obj)}")
+
+
+def UnionTypes(*types):
+    """Use `functools.reduce` to simulate `Union[*allowed_types]` on older
+    Python versions.
+    """
+    return reduce(lambda a, b: Union[a, b], types)
 
 
-def optionally_keyed_by(*arguments):
+def optionally_keyed_by(*arguments, use_msgspec=False):
```

Comment:
Oh, I see what you meant now by having separate functions for this.. For some reason I was thinking there'd be an object passed in here we could test, but that obviously isn't the case. I like this approach though.